crates/wasmtime/src/func/typed.rs

```diff
@@ -264,19 +292,20 @@ unsafe impl WasmTy for Option<Func> {
 ///
 /// This is implemented for bare types that can be passed to wasm as well as
 /// tuples of those types.
-pub unsafe trait WasmParams {
+pub unsafe trait WasmParams: Send {
+    #[doc(hidden)]
+    type Abi: Copy;
     #[doc(hidden)]
     fn typecheck(params: impl ExactSizeIterator<Item = crate::ValType>) -> Result<()>;
     #[doc(hidden)]
-    fn compatible_with_store(&self, store: &Store) -> bool;
+    fn into_abi(self, store: &mut StoreOpaque) -> Option<Self::Abi>;
```

Comment:
It's worth pointing out that I shifted functionality around here a bit. Now `into_abi` bakes in the `compatible_with_store` check as well as `invoke` takes/returns "ABI things" instead of raw values. This makes some more captures/closures `Copy` in the above `_call` function, but also notably means that the `invoke` function no longer needs a borrow on `Store` somewhow, which was much more difficult to provide.

Now the borrow is only required on `into_abi` here, then the call can happen without an explicit borrow on the store, and then re-ification back into actual values can happen after the call finishes.